### PR TITLE
feat(sb): add DataToBoardAdvisor to SmartBoard

### PR DESCRIPTION
**Pull request by ArranZeyuWang**
### PR includes

- [ ] fixed #0
- [ ] add / modify test cases
- [ ] documents, demos

Support i) input data directly and ii) integrate CA+LI process

### CODE REVIEW

File: packages/smart-board/src/SmartBoard/DataToBoard.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  private _charts: Chart[];
+
+  private _board: SmartBoard;
+
+  private _SBType: SmartBoardType;
+
+  constructor(props: DataToBoardProps) {
+    const { inputData, measures, dimensions, insightTypeList, insightNumber } = props;
+    // if there is empty data, return error message
+    if (!inputData.length) {
```
Comment (BBSQQ) [inline packages/smart-board/src/SmartBoard/DataToBoard.ts:28]: empty data don't need throw error

File: packages/smart-board/src/SmartBoard/DataToBoard.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+      throw new Error('The input dataset is empty!');
+    }
+
+    // compute insights by LI
+    const insightResult = getDataInsights(inputData, {
```
Comment (BBSQQ) [inline packages/smart-board/src/SmartBoard/DataToBoard.ts:33]: `getInsights` can be private method

File: packages/smart-board/src/SmartBoard/DataToBoard.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        measures: insightMeasures?.map((item) => item.field),
+        chartType: g2plotConfig?.chartType,
+        chartSchema: g2plotConfig,
+        description: insightCaption,
+      };
+      resultCharts.push(curChart);
+    });
+
+    // recommend charts by CA
+    // only if cannot compute insights by LI
```
Comment (BBSQQ) [inline packages/smart-board/src/SmartBoard/DataToBoard.ts:70]: private method

File: packages/smart-board/src/SmartBoard/DataToBoard.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
+/* eslint-disable no-underscore-dangle */
+import { statistics } from '@antv/data-wizard';
+import { getDataInsights } from '@antv/lite-insight';
+import { Advisor } from '@antv/chart-advisor';
+
+import { checkInsightTypes } from '../utils';
+import { INSIGHT_TYPES, G2PLOT_TYPES } from '../const';
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/DataToBoard.ts:7]: const -> constants ?

File: packages/smart-board/src/SmartBoard/DataToBoard.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+import { SmartBoard } from './ChartToBoard';
+
+import type { Chart, InputChart, SmartBoardType, DataToBoardProps } from '../interfaces';
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/DataToBoard.ts:11]: interfaces to types?

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+ * ChartToBoard as SmartBoard (the original SB)
+ * Get Dashboard from chart lists
+ * @public
+ */
+export class SmartBoard {
+  private charts: Chart[];
+
+  private _chartGraph: ChartGraph | undefined;
+
+  private _SBType: SmartBoardType;
```
Comment (chenluli) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:19]: 为什么都是 private 的成员，命名有的加 _ 有的不加

Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:19]: Why is the variable capitalized

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+/**
+ * ChartToBoard as SmartBoard (the original SB)
+ * Get Dashboard from chart lists
+ * @public
+ */
+export class SmartBoard {
+  private charts: Chart[];
+
+  private _chartGraph: ChartGraph | undefined;
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:17]: In my impression, the conclusion of one discussion is that private variables do not need to be prefixed

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+    if (!chart.id) {
+      addUuid(chart);
+    }
+    const initializedChart: Chart = Object.assign(chart);
+    return initializedChart;
+  }
+
+  constructor(inputCharts: InputChart[]) {
+    // if there are duplicate chart IDs, return error message
+    const chartIDs = inputCharts.map((d) => d.id).filter((d) => d !== undefined);
```
Comment (chenluli) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:39]: 命名上我们推荐 chartIds 这种形式？

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+   * - transform type from 'InputChart' to 'Chart'.
+   * @param chart
+   * @returns
+   */
+  private initChart(chart: InputChart) {
+    // if there is no chart ID, add it using uuid4
+    if (!chart.id) {
+      addUuid(chart);
+    }
+    const initializedChart: Chart = Object.assign(chart);
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:33]: Why use Object.assign? Do you want deep clone?

File: packages/smart-board/src/utils/calculator.ts [hunk 10]
```diff
@@ -110,3 +111,18 @@ export function getDegreeImportance(ChartList: Chart[]): any {
   });
   return degreeOfchart;
 }
+
+/**
+ * Calculate if the user-defined insight types accomplish with LI-defined
```
[the remark this answers is not in the document]
Comment (ArranZeyuWang) [reply packages/smart-board/src/utils/calculator.ts:116]: compared？

File: packages/smart-board/src/SmartBoard/DataToBoard.ts (hunk 1 above)
Comment (ArranZeyuWang) [reply packages/smart-board/src/SmartBoard/DataToBoard.ts:28]: get

File: packages/smart-board/src/SmartBoard/DataToBoard.ts (hunk 2 above)
Comment (ArranZeyuWang) [reply packages/smart-board/src/SmartBoard/DataToBoard.ts:33]: get

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts (hunk 8 above)
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:39]: chartIDs -> chartIds

File: packages/smart-board/src/interfaces.ts [hunk 11]
```diff
@@ -1,36 +1,41 @@
...
-  | 'association';
+  | 'change_point'
+  | 'time_series_outlier'
+  | 'majority'
+  | 'low_variance'
+  | 'correlation';
 
 /** Chart Type */
-type ChartType = 'line_chart' | 'column_chart' | 'pie_chart' | 'grouped_column_chart' | 'stack_column_chart';
+export type ChartType = 'line_chart' | 'column_chart' | 'pie_chart' | 'grouped_column_chart' | 'stack_column_chart';
```
[the remark this answers is not in the document]
Comment (ArranZeyuWang) [reply packages/smart-board/src/interfaces.ts:20]: Has been supported, it's an out-of-date definition, will refine later

File: packages/smart-board/src/interfaces.ts [hunk 12]
```diff
@@ -1,36 +1,41 @@
...
-type Subspace = {
+export type Subspace = {
   dimension: string;
   value: string;
 }[];
 
 /** Insight Type */
-type InsightType =
-  | 'outlier'
+export type InsightType =
```
[the remark this answers is not in the document]
Comment (ArranZeyuWang) [reply packages/smart-board/src/interfaces.ts:10]: > Does LI not export type?

No. There're also some problems in LI's measure definition, I'll refine the types of LI later.

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+    return initializedChart;
+  }
+
+  constructor(inputCharts: InputChart[]) {
+    // if there are duplicate chart IDs, return error message
+    const chartIDs = inputCharts.map((d) => d.id).filter((d) => d !== undefined);
+    const set = new Set(chartIDs);
+    if (set.size !== chartIDs.length) {
+      throw new Error('There are duplicate chart IDs, please check it.');
+    }
```
Comment (chenluli) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:43]: 这个需要很严格要求不能有重复么 感觉可以不阻塞后面流程

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  /**
+   * Get SB type
+   */
+  get type() {
+    if (!this._SBType) {
+      return null;
+    }
+    return this._SBType;
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:79]: u can replace these codes with `? : `

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  /**
+   * Get chart graph
+   */
+  get chartGraph() {
+    if (!this._chartGraph) {
+      this.generateChartGraph();
+    }
+    return this._chartGraph;
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:89]: u can replace these codes with `? : `

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+  /**
+   * Calculate the chart graph
+   * The graph structure is defined with reference to G6
+   * @param ChartListInfo
+   * @returns
+   */
+  private generateChartGraph() {
+    const connection = getChartConnection(this.charts);
+    const links: Link[] = [];
+    connection.forEach((d: any) => {
```
Comment (chenluli) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:101]: connection 有类型吗，为什么用 any

File: packages/smart-board/src/SmartBoard/DataToBoard.ts [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+
+/**
+ * DataToBoard as SmartBoard (new SB)
+ * Get Dashboard from input data
+ * @public
+ */
+export class DataToBoard {
```
[the remark this answers is not in the document]
Comment (ArranZeyuWang) [reply packages/smart-board/src/SmartBoard/DataToBoard.ts:18]: > DataToBoard 听着不像一个 class，起码主语得是个名词或名词短语

同意，需要优化

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+    // get chart order base on chartGraph
+    if (type === 'byCluster') {
+      if (!this._chartGraph) {
+        this.generateChartGraph();
+      }
+      const cluster = this.chartCluster();
+      const clusterIDSet = new Set(Object.values(cluster));
+      const degreeOfchart = getDegreeImportance(this.charts);
+      const a = 0.5;
+      const b = 0.5;
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:150]: Why are there 0.5? Can u add some comments?

File: packages/smart-board/src/SmartBoard/DataToBoard.ts [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+  private _charts: Chart[];
+
+  private _board: SmartBoard;
+
+  private _SBType: SmartBoardType;
```
Comment (Esorakouki) [inline packages/smart-board/src/SmartBoard/DataToBoard.ts:23]: is 'CHART ' included in _SBType of DataToBoard

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+      const clusterChartSet = this.charts.map((d) => {
+        return {
+          id: d.id,
+          clusterID: cluster[d.id],
+          computedScore: (d.score || 0) * a + (degreeOfchart.get(d.id) / (this.charts.length - 1)) * b,
+        };
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:156]: Why subtract 1

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+      });
+      const clusterChartGroupWithScore = clusterChartGroup
+        .map((d) => {
+          return {
+            clusterID: d[0].clusterID,
+            avgScore: d.map((chart) => chart.computedScore).reduce((a, b) => a + b) / d.length,
+            charts: d.sort((a, b) => b.computedScore - a.computedScore), // sort the charts inside the cluster
+          };
+        })
+        .sort((a, b) => b.avgScore - a.avgScore); // sort with the avgScore of each cluster
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:170]: replace a/b with some meaningful words

File: packages/smart-board/src/SmartBoard/DataToBoard.ts (hunk 19 above)
Comment (ArranZeyuWang) [reply packages/smart-board/src/SmartBoard/DataToBoard.ts:23]: > is 'CHART ' included in _SBType of DataToBoard

'chart' will be used only if user upload charts directly.

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+  }
+
+  /**
+   * Get chart cluster using the Louvain algorithm (https://www.npmjs.com/package/louvain-algorithm)
+   * @param chartsID
+   * @param ChartGraph
+   * @returns
+   */
+  chartCluster(): ChartCluster {
+    const chartsID = this.charts.map((c) => c.id);
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:193]: chartsID -> chartsId

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+    if (!this._chartGraph) {
+      this.generateChartGraph();
+    }
+    const MIN_DIFF_SHRESHOLD = 0.0000001;
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:197]: move the const to const file

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+    };
+    this._chartGraph = _chartGraph;
+  }
+
+  /**
+   * Calculate the chart score considering the insight score and the node degree importance and order them
+   * @param ChartListInfo
+   * @returns
+   */
+  chartOrder(type?: 'byInsightScore' | 'byCluster'): ChartOrder {
```
Comment (chenluli) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:123]: chartOrder 这里建议计算方法可以抽到外面文件里，因为之后可能会比较长？另外也许可以支持传入自定义函数，方便扩展不同的排序机制

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts (hunk 18 above)
Comment (chenluli) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:150]: 常数需要更有语义的命名，或者变成真常量（大写，单独维护）

File: packages/smart-board/src/SmartBoard/DataToBoard.ts [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+
+import type { Chart, InputChart, SmartBoardType, DataToBoardProps } from '../interfaces';
+
+/**
+ * DataToBoard as SmartBoard (new SB)
+ * Get Dashboard from input data
+ * @public
+ */
+export class DataToBoard {
+  private _charts: Chart[];
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/DataToBoard.ts:19]: In my impression, the conclusion of one discussion is that private variables do not need to be prefixed

File: packages/smart-board/src/SmartBoard/DataToBoard.ts (hunk 19 above)
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/DataToBoard.ts:23]: Why is the variable capitalized

File: packages/smart-board/src/SmartBoard/DataToBoard.ts [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+
+  /**
+   * Get SmartBoard
+   */
+  get board() {
+    if (!this._board) {
+      return null;
+    }
+    return this._board;
+  }
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/DataToBoard.ts:128]: u can replace these codes with `? :`
`get charts` `get type` `get board`

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts (hunk 24 above)
Comment (ArranZeyuWang) [reply packages/smart-board/src/SmartBoard/ChartToBoard.ts:123]: 后面会用新的排序算法 & 拆成新的文件；排序暂时没有支持的规划，权重可以支持自定义

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts (hunk 18 above)
Comment (ArranZeyuWang) [reply packages/smart-board/src/SmartBoard/ChartToBoard.ts:150]: get

File: packages/smart-board/src/SmartBoard/DataToBoard.ts (hunk 19 above)
Comment (ArranZeyuWang) [reply packages/smart-board/src/SmartBoard/DataToBoard.ts:23]: > but there is no entrance to input charts

The corresponding interface (for type "CHART") is on the file "ChartToBoard"

File: packages/smart-board/src/SmartBoard/DataToBoard.ts (hunk 3 above)
Comment (ArranZeyuWang) [reply packages/smart-board/src/SmartBoard/DataToBoard.ts:70]: done

File: packages/smart-board/src/SmartBoard/ChartToBoard.ts [hunk 27]
```diff
@@ -0,0 +1,204 @@
...
+
+  constructor(inputCharts: InputChart[]) {
+    // if there are duplicate chart IDs, return error message
+    const chartIds = inputCharts?.map((d) => d?.id).filter((d) => d !== undefined);
+    const set = new Set(chartIds);
+    if (set?.size !== chartIds?.length) {
+      throw new Error('There are duplicate chart IDs, please check it.');
+    }
+    // init charts
+    const initializedCharts = inputCharts.map((chart) => {
```
Comment (pddpd) [inline packages/smart-board/src/SmartBoard/ChartToBoard.ts:48]: add the optional chain？

**Review by pddpd — APPROVED** [review]
LGTM

**Review by pddpd — APPROVED** [review]

**Pull request merged**